### PR TITLE
feat: support subgraphs on Filecoin [SF-1156]

### PR DESCRIPTION
**Pull request by biga816**

### CODE REVIEW

**Review by leeshuzheng — APPROVED** [review]

File: src/utils/env.ts [hunk 1]
```diff
@@ -86,3 +86,11 @@ export const getReferralHashtags = (): string => {
     assert(referralHashtags, 'NEXT_PUBLIC_REFERRAL_HASHTAGS is not set');
     return referralHashtags;
 };
+
+export const getSubgraphUrl = (chainId: number): string | undefined => {
+    if (chainId === 314) {
+        return process.env.NEXT_PUBLIC_SUBGRAPH_URL_314;
```
Comment (as472) [inline src/utils/env.ts:92]: it does not assert the case when this env variable is missing.

File: src/utils/tests/env.test.ts [hunk 2]
```diff
@@ -184,3 +185,23 @@ describe('getReferralHashtags', () => {
...
+
+describe('getSubgraphUrl', () => {
+    it('should return the value of the environment variable', () => {
+        process.env.NEXT_PUBLIC_SUBGRAPH_URL_314 = 'test';
+        const subgraphUrl = getSubgraphUrl(314);
+        expect(subgraphUrl).toBe('test');
+        expect(typeof subgraphUrl).toBe('string');
+    });
+
+    it('should return empty if variable is not set', () => {
```
Comment (as472) [inline src/utils/tests/env.test.ts:197]: An empty string also means the variable is set.

File: src/pages/_app.tsx [hunk 3]
```diff
@@ -153,15 +154,20 @@ function App({ Component, pageProps }: AppProps) {
...
     const client = new ApolloClient({
         link: ApolloLink.split(
             operation => operation.getContext().type === 'point-dashboard',
             authLink.concat(httpLink),
-            new GraphApolloClient({ network: currentNetwork }).link
+            getSubgraphUrl(chainId)
+                ? createHttpLink({
+                      uri: getSubgraphUrl(chainId),
+                  })
+                : new GraphApolloClient({ network }).link
```
[the remark this answers is not in the document]
Comment (as472) [reply src/pages/_app.tsx:170]: 1. do we have a testnet filecoin subgraph as well?
2. Are you going to modify NEXT_PUBLIC_NON_SUBGRAPH_SUPPORTED_CHAIN_IDS?

File: src/pages/_app.tsx [hunk 4]
```diff
@@ -153,15 +154,20 @@ function App({ Component, pageProps }: AppProps) {
...
+        network: selectNetworkName(state),
+        chainId: state.blockchain.chainId,
+    }));
 
     const client = new ApolloClient({
         link: ApolloLink.split(
             operation => operation.getContext().type === 'point-dashboard',
             authLink.concat(httpLink),
-            new GraphApolloClient({ network: currentNetwork }).link
+            getSubgraphUrl(chainId)
```
Comment (as472) [inline src/pages/_app.tsx:166]: why call getSubgraphUrl(chainId) twice?

File: src/pages/_app.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (as472) [reply src/pages/_app.tsx:170]: 3. How do we test this is working?

**Comment by as472** [comment]
@biga816 it still connects to sepolia?
image

**Review by as472 — CHANGES_REQUESTED** [review]
Requesting addressal of the comments

File: src/utils/env.ts (hunk 1 above)
Comment (biga816) [reply src/utils/env.ts:92]: That's intentional. In the dev or stg environments, this variable is not set. So `undefined` should be returned when this env variable is missing.

File: src/pages/_app.tsx (hunk 4 above)
Comment (biga816) [reply src/pages/_app.tsx]: No reason. I fixed that.

File: src/utils/tests/env.test.ts (hunk 2 above)
Comment (biga816) [reply src/utils/tests/env.test.ts:197]: I fixed the test.

File: src/pages/_app.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (biga816) [reply src/pages/_app.tsx]: @as472 
> 1. do we have a testnet filecoin subgraph as well?

No, we don't have. This is deployed manually and hosted by Laconic freely.
We might have a chance to ask them to host it for our testnet use later.

> 2. Are you going to modify NEXT_PUBLIC_NON_SUBGRAPH_SUPPORTED_CHAIN_IDS?

Yes, we can remove 314 after NEXT_PUBLIC_SUBGRAPH_URL_314 is set.

> 3. How do we test this is working?

You can test it on your local environment or the UAT environment.

**Comment by biga816** [comment]
> @biga816 it still connects to sepolia?

Yes because we can use Laconic's subgraphs only on Filecion Mainnet.

**Comment by as472** [comment]
@biga816 looks good to me!
1. Please rebase with develop before merging as I can see some changes not part of develop anymore present in this PR.
2. Please trigger a UAT deployment so that we can test it out.

**Comment by biga816** [comment]
@as472 
> 1. Please rebase with develop before merging as I can see some changes not part of develop anymore present in this PR.

I have already done that! Could you approve this PR?

> 2. Please trigger a UAT deployment so that we can test it out.

Sure!

**Review by as472 — APPROVED** [review]

**Pull request merged**